### PR TITLE
Add error handling 

### PR DESCRIPTION
**Pull request by jatkinson1000**
This PR intends to improve the handling of errors in C++ to make things a little more instructive for users when things go wrong, rather than just having a bare stacktrace.

Currently there is some handling within FTorch to check tensors are created.
This adds handling on the C++ side to try and tell users why something has failed.

It is unclear how to test for this with pFUnit. At the moment I write some code, break it accordingly, and check I see the error expected...

Implemented:

- [x] generic error handling routines to be used throughout `ctorch.cpp`
    - These have not been made public or added to `ctorch.h` as designed to be used internally
- [x] generic routines to check tensors:
  - tensor is defined
  - tensor has requires_grad
  - gradient is defined for a tensor
- [x] usage of these throughout `ctorch.cpp`

Checking is not performed for getters (interrogation functions) or overloads (other than assign) as I felt probably unnecessary here.

This will introduce some overheads, but probably useful to have in for now as not expected to be significant and will help users in the development stage. Could consider a build with error handling off based on preprocessor flags in future if really desired.

### CODE REVIEW

**Comment by jatkinson1000** [comment]
In https://github.com/Cambridge-ICCS/FTorch/pull/347/commits/e9d7e5ea4259ddc7e6c46fa9a817cecc8f339f3a I abstract out the validation code into their own routines for re-use.

Since we do not expect to use them externally I have not added them to `ctorch.h`.

One thing that occurs to me is that when checking gradient is defined we effectively run the grad function in c++ twice (once to check, once to extract): https://github.com/Cambridge-ICCS/FTorch/blob/e9d7e5ea4259ddc7e6c46fa9a817cecc8f339f3a/src/ctorch.cpp#L160-L165

I wonder if we remove this check and replace it with a call to `﻿﻿validate_tensor` after we extract the gradient.
Doubt this is the serious bottleneck for now.
Any thoughts @jwallwork23

Side note, looks like pFUnit is having issues with CMake in GH actions 😢

**Comment by jwallwork23** [comment]
> I wonder if we remove this check and replace it with a call to `﻿﻿validate_tensor` after we extract the gradient. Doubt this is the serious bottleneck for now. One option would be to make the routine return the gradient, though I am wary of doing this with the pointer abstractions we are using. Any thoughts @jwallwork23

Yeah it might be better to take that approach, although to be sure we'd have to check if the Torch `grad` method is just a getter or if it actually computes anything. If it's just a getter than it probably doesn't matter.

> Side note, looks like pFUnit is having issues with CMake in GH actions 😢

Alas!

**Comment by jatkinson1000** [comment]
Rebased on main

**Comment by jatkinson1000** [comment]
Decided not to add tensor validation to the getters as this felt like overkill, but could if desired.
Similarly have not added checks to any overloads other than assignment.

Think this is now ready for review.

File: src/ctorch.cpp [hunk 1]
```diff
@@ -12,6 +12,25 @@
 #define GPU_DEVICE GPU_DEVICE_NONE
 #endif
 
+// =============================================================================
+// --- Functions to aid in consistent error handling
+// =============================================================================
+
+// Accept a string message and handle as error or warning if specified.
+// Can also accept a cleanup function if required.
```
Comment (jwallwork23) [inline src/ctorch.cpp:20]: Worth noting that this is only for the error case.

File: src/ctorch.cpp [hunk 2]
```diff
@@ -32,25 +50,23 @@ constexpr auto get_libtorch_dtype(torch_data_t dtype) {
...
+    ctorch_error("float16 not supported in Fortran");
     // See https://gcc.gnu.org/onlinedocs/gfortran/ISO_005fFORTRAN_005fENV.html
-    exit(EXIT_FAILURE);
   case torch_kFloat32:
     return torch::kFloat32;
   case torch_kFloat64:
     return torch::kFloat64;
   default:
-    std::cerr << "[WARNING]: unknown data type, setting to torch_kFloat32" << std::endl;
+    ctorch_error("unknown data type, setting to torch_kFloat32", true);
```
Comment (jwallwork23) [inline src/ctorch.cpp:60]: I find calling `ctorch_error` to raise a warning a bit unclear. Given that the function has different implementations for the error and warning cases, would it not make more sense to just have separate `ctorch_error` and `ctorch_warning` functions? That'd also make it clearer because currently the `cleanup` function gets silently ignored if you are just raising a warning.

File: src/ctorch.cpp [hunk 3]
```diff
@@ -147,24 +200,19 @@ torch_tensor_t torch_empty(int ndim, const int64_t *shape, torch_data_t dtype,
                            torch_device_t device_type, int device_index = -1,
                            const bool requires_grad = false) {
   torch::AutoGradMode enable_grad(requires_grad);
-  torch::Tensor *tensor = nullptr;
+  auto tensor = new torch::Tensor;
```
Comment (jwallwork23) [inline src/ctorch.cpp:203]: Good catch with these.

File: src/ctorch.cpp [hunk 4]
```diff
@@ -147,24 +200,19 @@ torch_tensor_t torch_empty(int ndim, const int64_t *shape, torch_data_t dtype,
...
                        .dtype(get_libtorch_dtype(dtype))
                        .device(get_libtorch_device(device_type, device_index))
                        .requires_grad(requires_grad);
-    tensor = new torch::Tensor;
     *tensor = torch::empty(vshape, options);
   } catch (const torch::Error &e) {
-    std::cerr << "[ERROR]: " << e.msg() << std::endl;
-    delete tensor;
-    exit(EXIT_FAILURE);
+    ctorch_error(e.msg(), false, [&]() { delete tensor; });
```
Comment (jwallwork23) [inline src/ctorch.cpp:213]: Neat use of lambdas!

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
This is great, thanks @jatkinson1000. My only request is that you separate the warning functionality out of `ctorch_error` as a separate `ctorch_warning` function.

File: src/ctorch.cpp (hunk 1 above)
Comment (jatkinson1000) [reply src/ctorch.cpp]: Amended in https://github.com/Cambridge-ICCS/FTorch/pull/347/commits/479d39ec0e82b938b16770f2c0193e7deb9e4be8 addressing below.

File: src/ctorch.cpp (hunk 2 above)
Comment (jatkinson1000) [reply src/ctorch.cpp]: Yes, that's fair and I did wonder about doing this as the function developed.
Should be covered in https://github.com/Cambridge-ICCS/FTorch/pull/347/commits/479d39ec0e82b938b16770f2c0193e7deb9e4be8

File: src/ctorch.cpp (hunk 4 above)
Comment (jatkinson1000) [reply src/ctorch.cpp]: If I'm honest, this is not 100% my code ;)
Though I learnt some things, perhaps.

**Comment by jatkinson1000** [comment]
Only final thing to note is that there are a couple of messages I couldn't handle with the new functions as they also output the `Torch::dtype` which I was not able to cast to string easily. For now I have left these as they were, though someone with better C++ may have a bright idea.

**Review by jwallwork23 — APPROVED** [review]
> Only final thing to note is that there are a couple of messages I couldn't handle with the new functions as they also output the `Torch::dtype` which I was not able to cast to string easily. For now I have left these as they were, though someone with better C++ may have a bright idea.

I think that's fine, we don't need to standardise everything - these changes already reduce multiplicity significantly.

**Review by niccolozanotti — APPROVED** [review]
Thanks @jatkinson1000 this is super useful.
